atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTopologyValidator.java

```diff
@@ -123,33 +123,42 @@ public Set<CassandraServer> getNewHostsWithInconsistentTopologies(
-                SafeArg.of("newlyAddedHosts", newlyAddedHosts),
-                SafeArg.of("allHosts", allHosts));
+                SafeArg.of("newlyAddedHosts", CassandraLogHelper.collectionOfHosts(newlyAddedHosts)),
+                SafeArg.of("allHosts", CassandraLogHelper.collectionOfHosts(allHosts.keySet())));
 
-        Map<CassandraServer, CassandraClientPoolingContainer> currentServers =
-                EntryStream.of(allHosts).removeKeys(newlyAddedHosts::contains).toMap();
-        Map<CassandraServer, CassandraClientPoolingContainer> newServers =
-                EntryStream.of(allHosts).filterKeys(newlyAddedHosts::contains).toMap();
+        Map<CassandraServer, HostIdResult> hostIdsByServerWithoutSoftFailures =
+                fetchHostIdsIgnoringSoftFailures(allHosts);
 
-        // This means we've not added any hosts yet.
-        if (currentServers.isEmpty()) {
-            return maybeGetConsistentClusterTopology(newServers)
+        Map<CassandraServer, HostIdResult> currentServersWithoutSoftFailures = EntryStream.of(
+                        hostIdsByServerWithoutSoftFailures)
+                .removeKeys(newlyAddedHosts::contains)
+                .toMap();
+        Map<CassandraServer, HostIdResult> newServersWithoutSoftFailures = EntryStream.of(
+                        hostIdsByServerWithoutSoftFailures)
+                .filterKeys(newlyAddedHosts::contains)
+                .toMap();
+
+        // This means currently we've no servers, therefore we need to come to a consensus on the new servers
+        if (currentServersWithoutSoftFailures.isEmpty()) {
+            return maybeGetConsistentClusterTopology(newServersWithoutSoftFailures)
                     .<Set<CassandraServer>>map(topology ->
                             // Do not add new servers which were unreachable
-                            Sets.difference(newlyAddedHosts, topology.serversInConsensus()))
-                    .orElse(newlyAddedHosts);
+                            Sets.difference(newServersWithoutSoftFailures.keySet(), topology.serversInConsensus()))
+                    .orElseGet(newServersWithoutSoftFailures::keySet);
         }
 
         // If a consensus can be reached from the current servers,
         // filter all new servers which have the same set of host ids.
         // Otherwise, if we cannot come to consensus on the current topology, just refuse to add any new servers.
-        return maybeGetConsistentClusterTopology(currentServers)
-                .map(topology -> EntryStream.of(fetchHostIdsIgnoringSoftFailures(newServers))
-                        .mapValues(HostIdResult::hostIds)
-                        .removeValues(topology.hostIds()::equals)
+        return maybeGetConsistentClusterTopology(currentServersWithoutSoftFailures)
+                .map(topology -> EntryStream.of(newServersWithoutSoftFailures)
+                        .removeValues(result -> result instanceof HostIdResult.SuccessfulHostIdResult
```

Comment:
I think I actually prefer the old way over casting if we're deciding against visitors (sorry!)